specification.json

```diff
@@ -1,171 +1,171 @@
             "RFC 2119 keyword": "MUST",
             "children": []
         },
         {
-            "id": "Requirement 1.14",
-            "machine_id": "requirement_1_14",
+            "id": "Requirement 1.4.6",
+            "machine_id": "requirement_1_4_6",
             "content": "In cases of normal execution, the `evaluation details` structure's `reason` field MUST contain the value of the `reason` field in the `flag resolution` structure returned by the configured `provider`, if the field is set.",
             "RFC 2119 keyword": "MUST",
             "children": []
         },
         {
-            "id": "Requirement 1.15",
-            "machine_id": "requirement_1_15",
+            "id": "Requirement 1.4.7",
+            "machine_id": "requirement_1_4_7",
             "content": "In cases of abnormal execution, the `evaluation details` structure's `error code` field MUST identify an error occurred during flag evaluation, having possible values `\"PROVIDER_NOT_READY\"`, `\"FLAG_NOT_FOUND\"`, `\"PARSE_ERROR\"`, `\"TYPE_MISMATCH\"`, or `\"GENERAL\"`.",
             "RFC 2119 keyword": "MUST",
             "children": []
         },
         {
-            "id": "Requirement 1.16",
-            "machine_id": "requirement_1_16",
+            "id": "Requirement 1.4.8",
+            "machine_id": "requirement_1_4_8",
             "content": "In cases of abnormal execution (network failure, unhandled error, etc) the `reason` field in the `evaluation details` SHOULD indicate an error.",
             "RFC 2119 keyword": "SHOULD",
             "children": []
         },
         {
-            "id": "Requirement 1.17",
-            "machine_id": "requirement_1_17",
-            "content": "The `evaluation options` structure's `hooks` field denotes a collection of hooks that the client MUST execute for the respective flag evaluation, in addition to those already configured.",
-            "RFC 2119 keyword": "MUST",
-            "children": []
-        },
-        {
-            "id": "Requirement 1.18",
-            "machine_id": "requirement_1_18",
-            "content": "Methods, functions, or operations on the client MUST NOT throw exceptions, or otherwise abnormally terminate. Flag evaluation calls must always return the `default value` in the event of abnormal execution. Exceptions include functions or methods for the purposes for configuration or setup.",
+            "id": "Requirement 1.4.9",
+            "machine_id": "requirement_1_4_9",
+            "content": "Methods, functions, or operations on the client MUST NOT throw exceptions, or otherwise abnormally terminate. Flag evaluation calls must always return the `default value` in the event of abnormal execution. Exceptions include functions or methods for the purposes for configuration or setup, and user-defined code such as hooks.",
```

Comment:
> I know I keep waffling on this. This makes me a little uneasy. I'm trying to think through the API from a user's perspective and I think we're going to end up with {reference ugly code}

I'm guilty of waffling too. For now, I feel that at the very least, for simplicity sake, we need to treat finally and error the same (if we allow `finally` to throw, we allow `error` to throw).

>This would mean that users don't get a guarantee that error hooks run (e.g. if error hook 1 throws, then error hook 2 wouldn't get invoked).

This sort of thing already happens in before/after - if a before hook throws, no other before hooks are executed. I dont think it's too astonishing to assume the same thing would occur in the error handler, but we should define it if we go this way.

On the side of throwing, I would say that `error` and `finally` hooks will likely be simple and low-risk enough that they rarely throw. Hooks are also not "SDK-code" proper, but code the app-author adds, which I think we can think about a bit differently from the main SDK functionality (Cloudbees considers "user-space" code "different" from primary SDK code as well, and allows some configuration around the error behavior, see [here](https://docs.cloudbees.com/docs/cloudbees-feature-management-api/latest/api-reference/java-api#_setuserspaceunhandlederrorhandler). I would also recommend that they throw unchecked exceptions, so that in Java, for instance, the compiler won't force you to wrap API calls in a try/catch. Throwing also allows for intentional re-throwing from an error, which might be useful in debugging situations, for instance, when developing a provider.

Lastly, the hooks semantics basically mirror a try/catch. Re-throwing in a catch results in propagation of the exception back up through the stack, and the rest of the catch block not executing. This is analogous to the behavior this PR defines for hooks.

On the side of swallowing... This is objectively the "safer" route. It does also allow us to guarantee all `error` and `finally` hooks are executed, and app code will NEVER break, not even if the author writes a really broken hook. I'm not 100% this is the least-astonishing way to go, though.

Reply:
> This sort of thing already happens in before/after - if a before hook throws, no other before hooks are executed. I dont think it's too astonishing to assume the same thing would occur in the error handler, but we should define it if we go this way.

An argument could be made for either approach. However, I think it may be worth considering catching errors so other hooks are still executed. Imagine if you have a hook that modifies context but runs after a misbehaving hook, this would lead to unexpected behavior.

Reply:
Ah, that makes sense.

Reply:
>When we start pulling in libraries for hooks, I get less sure. Now I need to validate that the hook I'm taking a dependency on isn't going to throw and crash my application, returning 5xx to the user.

Yep, that's a fair point. We do expect some of these to come from third party / a contrib repository

Reply:
I had another thought - what if a `finally` involves closing some resource and and an error results in NOT closing it, and therefor a leak? I think at the moment, the safest path is to never throw... so swallow everything and run ALL the `error` and `finally` hooks. That's the path we can take now to have the least regrets later.

I think if you guys are in agreement, I will update the spec for this. One other relevant issue is https://github.com/open-feature/spec/issues/39. We would likely want to log an error on this interface eventually, even if for now we just print to stderr.

Reply:
Always run finally and error hooks. Always swallow the errors.  I feel good about that.

Reply:
OK, I updated the spec for this: https://github.com/open-feature/spec/pull/90/files#r887153737